jgrapht-core/src/main/java/org/jgrapht/alg/util/Extension.java

```diff
@@ -34,60 +34,127 @@
+     * @return Extension object
+     */
+    public E getInstance()
+    {
+        return extensionFactory.create();
+    }
+
+    /**
+     * For a given prototype t, this function returns t's extension. If no encapsulation/extension
+     * for t exists, a new one is created and returned.
+     *
+     * NOTE: JK - Do we need want/this? There is no equivalent set(T t). Furthermore, it gets quite ambiguous if in some cases
+     * an Extension is created without a prototype (createInstance()) and sometimes an Extension is created through the get(T t) function.
+     * Only Extentions associated with a prototype t will be stored by the Manager. This leads to some very hard to track behavior. At the
+     * very least, give the function a proper name such as getSingletonInstance (as in contrast to the createInstance function above.
+     *
+     * @param t prototype
+     * @return Extension of prototype
+     */
+    @Deprecated
     public E get(T t)
     {
-        if (extensions.containsKey(t)) {
-            return extensions.get(t);
+        if (prototypeToExtensionMap.containsKey(t)) {
+            return prototypeToExtensionMap.get(t);
         }
 
-        E x = createInstance();
-        extensions.put(t, x);
+        E x = getInstance();
+        prototypeToExtensionMap.put(t, x);
         return x;
     }
 
     /**
-     * Factory capable of producing given extension objects of the given
+     * Creates a new extension object for prototype t if no such object exists, returns the old one otherwise.
+     */
+    public E getSingletonInstance(T t)
+    {
+        if (prototypeToExtensionMap.containsKey(t)) {
+            return prototypeToExtensionMap.get(t);
+        }
+
+        E x = getInstance();
+        prototypeToExtensionMap.put(t, x);
+        return x;
+    }
+
+    /**
+     * Factory capable of producing extension objects of the given
      * class-type
      *
-     * @param <E> extension concept class-type
+     * @param <E> class-type of extension
+     *
+     * NOTE: JK - Why is this class a subclass of the Extension/ExtensionManager class, instead of a class on its own.
      */
     public interface ExtensionFactory<E>
     {
         E create();
     }
 
+    /**
+     * Comments/Description missing
+     *
+     * NOTE: JK - Why do we want or need this? The only function this would offer is that all extensions have some common
+     * super type. However, none of the code requires that E extends BaseExtension. Furthermore, I'm not sure why this
+     * class is a subclass of the Extension/ExtensionManager class, instead of a class on its own.
+     */
     public static abstract class BaseExtension
     {
         public BaseExtension()
         {
         }
     }
 
+    /**
+     * Comments/Description missing
+     *
+     * NOTE: JK - Do we really need this? This seems well beyond what any user would require. Better to keep the library as concise as possible.
+     */
```

Comment:
That thing has been exposed to manifest problems with construction of inheritors of the `ExtensionBase`. Though, since it stay unused, it could be killed safely.